lib/Dialect/ModArith/Conversions/ModArithToArith/ModArithToArith.cpp

```diff
@@ -250,6 +269,127 @@ struct ConvertMac : public OpConversionPattern<MacOp> {
+        auto modWidth = inputModArith.getModulus().getValue().getBitWidth();
+        auto rnsWidth = cast<ModArithType>(outputRNS.getBasisTypes()[0])
+                            .getModulus()
+                            .getValue()
+                            .getBitWidth();
+        auto modintegerType = IntegerType::get(op.getContext(), modWidth);
+        auto rnsIntegerType = IntegerType::get(op.getContext(), rnsWidth);
+        auto intConst = getInteger(&b, modintegerType);
+        std::vector<Value> rnsElements;
+        for (auto basisType : outputRNS.getBasisTypes()) {
+          auto modArithType = cast<ModArithType>(basisType);
+          auto modulus = modArithType.getModulus().getValue();
+          modulus = modulus.zext(modWidth);
+          auto cmod = intConst(modulus);
+          auto remu = b.create<arith::RemUIOp>(adaptor.getInput(), cmod);
+          auto trunci = b.create<arith::TruncIOp>(rnsIntegerType, remu);
+          rnsElements.push_back(trunci);
+        }
+        auto tensor = b.create<tensor::FromElementsOp>(ValueRange(rnsElements));
+        rewriter.replaceOp(op, tensor);
+      } else {
+        llvm_unreachable("Verifier should make sure this doesn't happen.");
+      }
+    } else if (auto inputRNS = dyn_cast<rns::RNSType>(inputType)) {
+      auto outputModArith = cast<ModArithType>(outputType);
+      auto bigModulus = outputModArith.getModulus().getValue();
+      auto outputWidth = bigModulus.getBitWidth();
+      // The width of each element in the RNS type.
+      auto rnsWidth = cast<ModArithType>(inputRNS.getBasisTypes()[0])
+                          .getModulus()
+                          .getValue()
+                          .getBitWidth();
+      // The width of intermediate results: In the formula (from OpenFHE)
+      // ```
+      // Sigma(i = 0 --> t-1) { M(i) * qt/qi * [(qt/qi)^(-1) mod qi] } mod qt
+      // ```
+      // The width of `qt/qi` is bounded by `outputWidth`. The width of `M(i)`
+      // is bounded by `rnsWidth`, and the value of `(qt/qi)^(-1) mod qi` is
+      // bounded by `qi`, so its width is bounded by `rnsWidth`.
+      // So the width of `M(i) * qt/qi * [(qt/qi)^(-1) mod qi]` is bounded by
+      // `outputWidth + 2 * rnsWidth`.
+      auto width = outputWidth + 2 * rnsWidth +
+                   llvm::Log2_64_Ceil(inputRNS.getBasisTypes().size());
+      bigModulus = bigModulus.zext(width);
+      auto integerType = IntegerType::get(op.getContext(), width);
+      auto intConst = getInteger(&b, integerType);
+      auto bigModulusConst = intConst(bigModulus);
+      Value sum = intConst(APInt::getZero(width));
+      for (int64_t i = 0; i < inputRNS.getBasisTypes().size(); i++) {
+        auto modArithType = cast<ModArithType>(inputRNS.getBasisTypes()[i]);
+        auto modulus = modArithType.getModulus().getValue();
+        modulus = modulus.zext(width);
+        auto tmp = bigModulus.udiv(modulus);
```

Comment:
It would help slightly to have the variable names match the formulas above, or to have a comment saying which is which. `bigModulus` = `qt` and `modulus` = `qi`